src/core/Akka.Streams/Dsl/Restart.cs

```diff
@@ -0,0 +1,490 @@
+            {
+                var sinkIn = CreateSubInlet(_stage.Out);
+                _stage.SourceFactory().RunWith(sinkIn.Sink, SubFusingMaterializer);
+                if (IsAvailable(_stage.Out))
+                {
+                    sinkIn.Pull();
+                }
+            }
+
+            protected override void Backoff()
+            {
+                SetHandler(_stage.Out, () =>
+                {
+                    // do nothing
+                });
+            }
+        }
+    }
+
+    /// <summary>
+    /// A RestartSink wraps a <see cref="Sink"/> that gets restarted when it completes or fails.
+    /// They are useful for graphs that need to run for longer than the <see cref="Sink"/> can necessarily guarantee it will, for
+    /// example, for <see cref="Sink"/> streams that depend on a remote server that may crash or become partitioned. The
+    /// RestartSink ensures that the graph can continue running while the <see cref="Sink"/> restarts.
+    /// </summary>
+    public static class RestartSink
+    {
+        /// <summary>
+        /// Wrap the given <see cref="Sink"/> with a <see cref="Sink"/> that will restart it when it fails or complete using an exponential
+        /// backoff.
+        /// This <see cref="Sink"/> will never cancel, since cancellation by the wrapped <see cref="Sink"/> is always handled by restarting it.
+        /// The wrapped <see cref="Sink"/> can however be completed by feeding a completion or error into this <see cref="Sink"/>. When that
+        /// happens, the <see cref="Sink"/>, if currently running, will terminate and will not be restarted. This can be triggered
+        /// simply by the upstream completing, or externally by introducing a <see cref="IKillSwitch"/> right before this <see cref="Sink"/> in the
+        /// graph.
+        /// The restart process is inherently lossy, since there is no coordination between cancelling and the sending of
+        /// messages. When the wrapped <see cref="Sink"/> does cancel, this <see cref="Sink"/> will backpressure, however any elements already
+        /// sent may have been lost.
+        /// This uses the same exponential backoff algorithm as <see cref="Akka.Pattern.Backoff"/>.
+        /// </summary>
+        /// <param name="sinkFactory">A factory for producing the <see cref="Sink"/> to wrap.</param>
+        /// <param name="minBackoff">Minimum (initial) duration until the child actor will started again, if it is terminated</param>
+        /// <param name="maxBackoff">The exponential back-off is capped to this duration</param>
+        /// <param name="randomFactor">After calculation of the exponential back-off an additional random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay. In order to skip this additional delay pass in `0`.</param>
+        public static Sink<T, NotUsed> WithBackoff<T>(Func<Sink<T, NotUsed>> sinkFactory, TimeSpan minBackoff, TimeSpan maxBackoff, double randomFactor) 
```

Comment:
`sinkFactory` should have a signature `Creator[Sink[T, _]]`, I don't know how to do it in C#